service/common/src/main/java/org/apache/polaris/service/events/listeners/FileBufferListingTask.java

```diff
@@ -0,0 +1,111 @@
+    private final int timeToFlush;
+
+    public FileBufferListingTask(String bufferDirectory, Function<TaskSubmissionInput, Future> taskSubmitter, ConcurrentHashMap<String, Future> activeFlushFutures, int timeToFlush, BiConsumer<String, Integer> rotateShard, BiConsumer<String, List<PolarisEvent>> eventWriter) {
+        this.bufferDirectory = bufferDirectory;
+        this.taskSubmitter = taskSubmitter;
+        this.activeFlushFutures = activeFlushFutures;
+        this.timeToFlush = timeToFlush;
+        this.rotateShard = rotateShard;
+        this.eventWriter = eventWriter;
+    }
+
+    @Override
+    public void run() {
+        LOGGER.info("Attempting to list all event buffer files");
+        try {
+            // Example, if `topLevelDirectory` is /var/tmp/...
+            File topLevelDirectory = new File(this.bufferDirectory);
+            File[] realmDirs = topLevelDirectory.listFiles();
+            if (realmDirs == null || realmDirs.length == 0) {
+                LOGGER.debug("No event buffer files found in path: {}", topLevelDirectory.getAbsolutePath());
+                return;
+            }
+            for (File realmDir : realmDirs) {
+                // Then `realmDir` would be something like /var/tmp/POLARIS/
+                String realmId = realmDir.getName();
+                LOGGER.debug("Buffer realm directory found for realmId: {}", realmId);
+                ArrayList<String> filesToFlush = new ArrayList<>();
+                File[] bufferDirs = realmDir.listFiles();
+                if (bufferDirs == null || bufferDirs.length == 0) {
+                    LOGGER.debug("No buffer directories found in path: {}", realmDir.getAbsolutePath());
+                    continue;
+                }
+                for (File bufferDir : bufferDirs) {
+                    // And `bufferDir` would be something like /var/tmp/POLARIS/polaris-event-buffer-shard-1/
+                    File[] bufferFiles = bufferDir.listFiles();
+                    if (bufferFiles == null || bufferFiles.length == 0) {
+                        LOGGER.debug("No buffer files found in path: {}", bufferDir.getAbsolutePath());
+                        continue;
+                    }
+                    // Get all buffer files except for the latest one
+                    File latestFile = Arrays.stream(bufferFiles).max(File::compareTo).orElse(null);
+                    Arrays.stream(bufferFiles).map(File::getAbsolutePath).filter(file -> !file.equals(latestFile.getAbsolutePath()) && !activeFlushFutures.containsKey(file)).forEach(filesToFlush::add);
+
+                    // Get the buffer start time from the file name
+                    String[] fileDeconstructed = latestFile.getName().split("-");
+                    // Check if this buffer has been alive for longer than the `timeToFlush`
+                    if (System.currentTimeMillis() - Long.parseLong(fileDeconstructed[fileDeconstructed.length - 1]) > timeToFlush) {
```

Comment:
Interesting be ware, people think System.currentTimeMillis is not reliable so this check is not reliable :P and events gets missed, orthogonal question what if server clock time changes ?

Reply:
I don't think we have another way out here either. System.nanoTime does not provide any guarantees across JVM restarts.